core/src/main/java/com/segment/analytics/kotlin/core/utilities/OkHttpURLConnection.kt

```diff
@@ -0,0 +1,393 @@
+import okhttp3.RequestBody.Companion.toRequestBody
+import okhttp3.Response
+import okio.Buffer
+import okio.GzipSink
+import okio.buffer
+import java.io.IOException
+import java.io.InputStream
+import java.io.OutputStream
+import java.net.HttpURLConnection
+import java.net.ProtocolException
+import java.net.URL
+import kotlin.io.bufferedReader
+import kotlin.io.readBytes
+
+internal class OkHttpURLConnection(
+    url: URL,
+    private val client: OkHttpClient
+) : HttpURLConnection(url) {
+
+    private var response: Response? = null
+    private var requestBodyBuffer: Buffer? = null
+    private var connected = false
+
+    private val requestBuilder = Request.Builder().url(url)
+    private val requestProperties = mutableMapOf<String, MutableList<String>>()
+
+    // OkHttp-only state management
+    private var _requestMethod = "GET"
+    private var _doInput = true
+    private var _doOutput = false
+    private var _allowUserInteraction = false
+    private var _useCaches = true
+    private var _ifModifiedSince = 0L
+    private var _connectTimeout = 15000
+    private var _readTimeout = 20000
+    private var _instanceFollowRedirects = true
+
+    override fun disconnect() {
+        response?.close()
+        connected = false
+    }
+
+    override fun usingProxy(): Boolean = false
+
+    @Throws(IOException::class)
+    override fun connect() {
+        if (connected) return
+
+        try {
+            val finalRequest = buildRequest()
+            response = client.newCall(finalRequest).execute()
+            responseCode = response!!.code
+            responseMessage = response!!.message
+            connected = true
+        } catch (e: Exception) {
+            throw IOException("Connection failed", e)
+        }
+    }
+
+    internal fun buildRequest(): Request {
+        val builder = requestBuilder
+
+        requestProperties.forEach { (key, values) ->
+            values.forEach { value ->
+                builder.addHeader(key, value)
+            }
+        }
+
+        when (_requestMethod) {
+            "GET" -> builder.get()
+            "POST" -> {
+                val body = requestBodyBuffer?.let { buffer ->
+                    val mediaType = getRequestProperty("Content-Type")?.toMediaType()
+                        ?: "text/plain".toMediaType()
+                    buffer.readByteArray().toRequestBody(mediaType)
+                } ?: "".toRequestBody("text/plain".toMediaType())
+                builder.post(body)
+            }
+            "PUT" -> {
+                val body = requestBodyBuffer?.readByteArray()?.toRequestBody(
+                    getRequestProperty("Content-Type")?.toMediaType() ?: "text/plain".toMediaType()
+                ) ?: "".toRequestBody("text/plain".toMediaType())
+                builder.put(body)
+            }
+            "DELETE" -> builder.delete()
+            "HEAD" -> builder.head()
+            else -> throw ProtocolException("Unknown method: $_requestMethod")
```

Comment:
With Enum does this go away?

Reply:
yeah. if it's enum, the else branch is not necessary. I could change it to enum, but the `setRequestMethod` takes a string. we'd have to do a conversion internally. besides that users are not going to use the enum, because the connection auto converts GET to POST. they are purely internal.